modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/TransportRankEvalAction.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Override
+    protected void doExecute(RankEvalRequest request, ActionListener<RankEvalResponse> listener) {
+        RankEvalSpec evaluationSpecification = request.getRankEvalSpec();
+        List<String> indices = evaluationSpecification.getIndices();
+
+        Collection<RatedRequest> ratedRequests = evaluationSpecification.getRatedRequests();
+        AtomicInteger responseCounter = new AtomicInteger(ratedRequests.size());
+        Map<String, EvalQueryQuality> partialResults = new ConcurrentHashMap<>(
+                ratedRequests.size());
+        Map<String, Exception> errors = new ConcurrentHashMap<>(ratedRequests.size());
+
+        Map<String, TemplateScript.Factory> scriptsWithoutParams = new HashMap<>();
+        for (Entry<String, Script> entry : evaluationSpecification.getTemplates().entrySet()) {
+            scriptsWithoutParams.put(entry.getKey(), scriptService.compile(entry.getValue(), TemplateScript.CONTEXT));
+        }
+
+        for (RatedRequest ratedRequest : ratedRequests) {
+            SearchSourceBuilder ratedSearchSource = ratedRequest.getTestRequest();
+            if (ratedSearchSource == null) {
+                Map<String, Object> params = ratedRequest.getParams();
+                String templateId = ratedRequest.getTemplateId();
+                TemplateScript.Factory templateScript = scriptsWithoutParams.get(templateId);
+                String resolvedRequest = templateScript.newInstance(params).execute();
+                try (XContentParser subParser = createParser(namedXContentRegistry, new BytesArray(resolvedRequest), XContentType.JSON)) {
+                    ratedSearchSource = SearchSourceBuilder.fromXContent(subParser);
+                } catch (IOException e) {
+                    listener.onFailure(e);
+                }
+            }
+
+            List<String> summaryFields = ratedRequest.getSummaryFields();
+            if (summaryFields.isEmpty()) {
+                ratedSearchSource.fetchSource(false);
+            } else {
+                ratedSearchSource.fetchSource(summaryFields.toArray(new String[summaryFields.size()]), new String[0]);
+            }
+
+            SearchRequest templatedRequest = new SearchRequest(indices.toArray(new String[indices.size()]), ratedSearchSource);
+            final RankEvalActionListener searchListener = new RankEvalActionListener(listener,
+                    evaluationSpecification.getMetric(), ratedRequest, partialResults, errors, responseCounter);
+            RequestTask task = new RequestTask(templatedRequest, searchListener);
+            taskQueue.add(task);
+        }
+
+        // Execute top n tasks, further execution is triggered in RankEvalActionListener
+        for (int i = 0; (i < Math.min(ratedRequests.size(),
+                evaluationSpecification.getMaxConcurrentSearches())); i++) {
+            RequestTask task = taskQueue.poll();
+            client.search(task.request, task.searchListener);
```

Comment:
Why not using multi search directly ? If you have a lot of requests you can split in multiple requests and relies on the concurrent execution of the multi search request.

Reply:
Good point, I think we just didn't think about this at the time. I will investigate how this would integrate with the response listeners we currently have.

Reply:
I opened https://github.com/elastic/elasticsearch/pull/27520 for a version that uses multi search. Let me know what you think. I opened it as a separate PR to make it easier to review and also so we can drop it if its not what is wanted, but I can also pull it in here with the rest if you prefer that.